src/cmd/ksh93/sh/arith.c

```diff
@@ -194,24 +194,30 @@ static Namval_t *scope(Namval_t *np,struct lval *lvalue,int assign)
 	return np;
 }
 
-static Math_f sh_mathstdfun(const char *fname, size_t fsize, short * nargs)
+/* lookup a function in the standard math function table */
+static Math_f sh_mathstdfun(const char *fname, size_t fsize, short *nargs)
 {
 	const struct mathtab *tp;
-	char c = fname[0];
-	for(tp=shtab_math; *tp->fname; tp++)
+	char firstc = fname[0];
+	/* first byte of tp->fname is num. args and return type, unless empty */
+	for(tp=shtab_math; tp->fname[0]=='\0'; tp++)
 	{
-		if(*tp->fname > c)
+		/* shtab_math is in alphabetic order - if first character is greater, we're done */
+		if(tp->fname[1] > firstc)
```

Comment:
This change is incorrect, because it assumes that shtab_math is in alphabetical order. Unfortunately, a look at the generated shtab_math in arch/*/src/cmd/ksh93/FEATURE/math shows that it is not sorted, alphabetically or otherwise.

What on earth AT&T meant to accomplish with the original `if(*tp->fname > c)` is a mystery. In practice, it's a no-op. The first byte of fname is non-printable and never greater than octal 12, i.e. 10 (going by the generated shtab_math on my machine). It is compared against a printable ASCII character which will always be greater than that, so the `break` is never reached. Those two lines can be deleted.